File: backend/app/core/annotations.py
"""The sender's own marks on a page: a pen drawing and a text box.

Both are ordinary ``Field`` rows of a new ``FieldType``, which is what lets them
ride the field endpoints, the bulk save, the geometry validation and the final
PDF overlay that already exist. What separates them from every other type is
that they are *authored* content rather than something a recipient does: they
are always ``read_only`` and never ``required``, so they place no obligation on
anybody and the signing surface draws them as inert marks.

The shape each one carries in ``Field.options`` is normalised here, in one
place, because three consumers read it: the API (on the way in), the PDF
overlay (on the way out) and the browser. Anything malformed is dropped rather
than rejected -- a stray key in an options blob is not worth failing a save the
sender cannot correct.

Drawing coordinates are stored *normalised* to the field's own box (0..1, with
a top-left origin, the same origin the field's ``x``/``y`` use). That is what
makes a drawing survive being dragged, resized, or drawn at one zoom level and
rendered at another: the stroke is defined relative to its box, not to the page.
"""

from __future__ import annotations

from typing import Any

from app.models.enums import FieldType

#: Field types that are the sender's mark on the page rather than a recipient's
#: obligation. Kept here (not in ``enums``) so the rules that go with them --
#: always read-only, never required -- live beside the shape they carry.
ANNOTATION_TYPES = frozenset({FieldType.drawing, FieldType.textbox})

#: The base-14 PDF families reportlab can draw without an embedded font file.
#: The key is what the client stores and the value is the reportlab base name;
#: bold/italic pick the matching face in ``pdf_font_name``.
TEXTBOX_FONTS: dict[str, str] = {
    "helvetica": "Helvetica",
    "times": "Times",
    "courier": "Courier",
}
DEFAULT_TEXTBOX_FONT = "helvetica"
DEFAULT_TEXTBOX_SIZE = 12.0
MIN_TEXTBOX_SIZE = 6.0
MAX_TEXTBOX_SIZE = 96.0
DEFAULT_INK = "#0f172a"

MIN_PEN_WIDTH = 0.5
MAX_PEN_WIDTH = 12.0
DEFAULT_PEN_WIDTH = 2.0
#: A single gesture is sampled on every pointermove, which on a trackpad is far
#: more points than the mark needs. Capping keeps one field's blob bounded.
MAX_STROKES = 200
MAX_POINTS_PER_STROKE = 4000


def is_annotation(field_type: Any) -> bool:
    return field_type in ANNOTATION_TYPES


def _color(value: Any, fallback: str = DEFAULT_INK) -> str:
    """A ``#rrggbb`` string, or the fallback. Never raises: colour is cosmetic."""
    if not isinstance(value, str):
        return fallback
    text = value.strip()
    if len(text) != 7 or not text.startswith("#"):
        return fallback
    try:
        int(text[1:], 16)
    except ValueError:
        return fallback
    return text.lower()


def _number(value: Any, *, low: float, high: float, fallback: float) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return fallback
    if number != number or number in (float("inf"), float("-inf")):  # NaN / inf
        return fallback
    return max(low, min(high, number))


def _unit(value: Any) -> float | None:
    """One drawing coordinate, clamped into the field's box."""
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    if number != number or number in (float("inf"), float("-inf")):
        return None
    return max(0.0, min(1.0, number))


def normalize_textbox_options(options: Any) -> dict[str, Any]:
    raw = options if isinstance(options, dict) else {}
    font = raw.get("font")
    font = font.lower() if isinstance(font, str) else ""
    return {
        "kind": "textbox",
        "font": font if font in TEXTBOX_FONTS else DEFAULT_TEXTBOX_FONT,
        "size": round(
            _number(raw.get("size"), low=MIN_TEXTBOX_SIZE, high=MAX_TEXTBOX_SIZE, fallback=DEFAULT_TEXTBOX_SIZE), 2
        ),
        "bold": bool(raw.get("bold")),
        "italic": bool(raw.get("italic")),
        "color": _color(raw.get("color")),
    }


def normalize_drawing_options(options: Any) -> dict[str, Any]:
    raw = options if isinstance(options, dict) else {}
    strokes: list[list[list[float]]] = []
    for stroke in (raw.get("strokes") or [])[:MAX_STROKES]:
        if not isinstance(stroke, (list, tuple)):
            continue
        points: list[list[float]] = []
        for point in stroke[:MAX_POINTS_PER_STROKE]:
            if not isinstance(point, (list, tuple)) or len(point) < 2:
                continue
            x, y = _unit(point[0]), _unit(point[1])
            if x is None or y is None:
                continue
            points.append([round(x, 4), round(y, 4)])
        # A one-point stroke is a dot; the renderer draws it as a round cap, so
        # it is kept. An empty one is not a mark at all.
        if points:
            strokes.append(points)
    return {
        "kind": "drawing",
        "color": _color(raw.get("color")),
        "stroke": round(
            _number(raw.get("stroke"), low=MIN_PEN_WIDTH, high=MAX_PEN_WIDTH, fallback=DEFAULT_PEN_WIDTH), 2
        ),
        "strokes": strokes,
    }


def normalize_options(field_type: Any, options: Any) -> Any:
    """Options as stored for ``field_type``; untouched for a non-annotation."""
    if field_type == FieldType.textbox:
        return normalize_textbox_options(options)
    if field_type == FieldType.drawing:
        return normalize_drawing_options(options)
    return options


def pdf_font_name(font: str, *, bold: bool = False, italic: bool = False) -> str:
    """The reportlab base-14 face for a stored family plus its style flags."""
    base = TEXTBOX_FONTS.get((font or "").lower(), TEXTBOX_FONTS[DEFAULT_TEXTBOX_FONT])
    if base == "Times":
        if bold and italic:
            return "Times-BoldItalic"
        if bold:
            return "Times-Bold"
        if italic:
            return "Times-Italic"
        return "Times-Roman"
    if bold and italic:
        return f"{base}-BoldOblique"
    if bold:
        return f"{base}-Bold"
    if italic:
        return f"{base}-Oblique"
    return base


def rgb(color: str) -> tuple[float, float, float]:
    """``#rrggbb`` as reportlab's 0..1 triple."""
    text = _color(color)
    return (int(text[1:3], 16) / 255, int(text[3:5], 16) / 255, int(text[5:7], 16) / 255)

File: frontend/lib/sf/annotations.ts
/**
 * Page annotations — the sender's own marks on the document: a freehand pen
 * drawing and a text box in a chosen face and size.
 *
 * This is the client half of `backend/app/core/annotations.py`; the shape below
 * is the same one the API normalises and the final PDF draws, so the two must
 * be changed together. Both are stored as ordinary fields of type `drawing` /
 * `textbox`, always read-only and never required: they place no obligation on a
 * recipient, they are simply drawn.
 *
 * Stroke points are normalised to the field's own box (0..1, top-left origin),
 * which is why a drawing survives being dragged, resized, or drawn at one zoom
 * level and rendered at another — the mark is defined relative to its box, not
 * to the page.
 */

export const ANNOTATION_TYPES = ['drawing', 'textbox'] as const;
export type AnnotationType = (typeof ANNOTATION_TYPES)[number];

export function isAnnotationType(type: string): type is AnnotationType {
  return type === 'drawing' || type === 'textbox';
}

/** The three base-14 families the PDF can draw without an embedded font. */
export const TEXTBOX_FONTS: { id: string; label: string; stack: string }[] = [
  { id: 'helvetica', label: 'Helvetica (sans-serif)', stack: 'Helvetica, Arial, var(--font-sans), sans-serif' },
  { id: 'times', label: 'Times (serif)', stack: '"Times New Roman", Times, Georgia, serif' },
  { id: 'courier', label: 'Courier (monospace)', stack: '"Courier New", Courier, var(--font-geist-mono), monospace' },
];
export const TEXTBOX_SIZES = [8, 9, 10, 11, 12, 14, 16, 18, 20, 24, 28, 32, 40, 48];
export const DEFAULT_TEXTBOX_FONT = 'helvetica';
export const DEFAULT_TEXTBOX_SIZE = 12;
export const MIN_TEXTBOX_SIZE = 6;
export const MAX_TEXTBOX_SIZE = 96;
export const DEFAULT_INK = '#0f172a';
export const INK_COLORS = ['#0f172a', '#1d4ed8', '#dc2626', '#047857', '#b45309'];
export const MIN_PEN_WIDTH = 0.5;
export const MAX_PEN_WIDTH = 12;
export const DEFAULT_PEN_WIDTH = 2;

export type TextboxOptions = {
  kind: 'textbox';
  font: string;
  size: number;
  bold: boolean;
  italic: boolean;
  color: string;
};

/** One stroke is a list of `[x, y]` pairs, each 0..1 within the field box. */
export type Stroke = [number, number][];

export type DrawingOptions = {
  kind: 'drawing';
  color: string;
  stroke: number;
  strokes: Stroke[];
};

type Raw = Record<string, unknown> | unknown[] | null | undefined;

function asRecord(options: Raw): Record<string, unknown> {
  return options && !Array.isArray(options) && typeof options === 'object'
    ? (options as Record<string, unknown>)
    : {};
}

function color(value: unknown, fallback = DEFAULT_INK): string {
  return typeof value === 'string' && /^#[0-9a-fA-F]{6}$/.test(value.trim())
    ? value.trim().toLowerCase()
    : fallback;
}

function clamp(value: unknown, low: number, high: number, fallback: number): number {
  const n = typeof value === 'number' ? value : parseFloat(String(value));
  if (!Number.isFinite(n)) return fallback;
  return Math.max(low, Math.min(high, n));
}

export function textboxOptions(options: Raw): TextboxOptions {
  const raw = asRecord(options);
  const font = typeof raw.font === 'string' ? raw.font.toLowerCase() : '';
  return {
    kind: 'textbox',
    font: TEXTBOX_FONTS.some(f => f.id === font) ? font : DEFAULT_TEXTBOX_FONT,
    size: Math.round(clamp(raw.size, MIN_TEXTBOX_SIZE, MAX_TEXTBOX_SIZE, DEFAULT_TEXTBOX_SIZE) * 100) / 100,
    bold: raw.bold === true,
    italic: raw.italic === true,
    color: color(raw.color),
  };
}

export function drawingOptions(options: Raw): DrawingOptions {
  const raw = asRecord(options);
  const source = Array.isArray(raw.strokes) ? raw.strokes : [];
  const strokes: Stroke[] = [];
  for (const entry of source) {
    if (!Array.isArray(entry)) continue;
    const points: Stroke = [];
    for (const point of entry) {
      if (!Array.isArray(point) || point.length < 2) continue;
      const x = clamp(point[0], 0, 1, NaN);
      const y = clamp(point[1], 0, 1, NaN);
      if (!Number.isFinite(x) || !Number.isFinite(y)) continue;
      points.push([x, y]);
    }
    if (points.length) strokes.push(points);
  }
  return {
    kind: 'drawing',
    color: color(raw.color),
    stroke: clamp(raw.stroke, MIN_PEN_WIDTH, MAX_PEN_WIDTH, DEFAULT_PEN_WIDTH),
    strokes,
  };
}

/** The CSS `font-family` for a stored family id. */
export function textboxFontStack(font: string): string {
  const entry = TEXTBOX_FONTS.find(f => f.id === (font || '').toLowerCase());
  return (entry ?? TEXTBOX_FONTS[0]).stack;
}

/**
 * One stroke as an SVG `d`, scaled into a box `w` × `h` (CSS px). A one-point
 * stroke is a tap: it is drawn as a zero-length line, which a round linecap
 * renders as the dot the sender made.
 */
export function strokePath(stroke: Stroke, w: number, h: number): string {
  if (!stroke.length) return '';
  const at = (p: [number, number]) => (p[0] * w).toFixed(2) + ' ' + (p[1] * h).toFixed(2);
  if (stroke.length === 1) return 'M' + at(stroke[0]) + 'L' + at(stroke[0]);
  return 'M' + at(stroke[0]) + stroke.slice(1).map(p => 'L' + at(p)).join('');
}

/**
 * The bounding box of a freehand gesture, in page points, padded so a stroke
 * drawn with a thick nib is not clipped by its own box — and never thinner than
 * a couple of points, so a straight horizontal line still has a box to live in.
 */
export function strokeBounds(points: [number, number][], pad: number) {
  const xs = points.map(p => p[0]);
  const ys = points.map(p => p[1]);
  const x = Math.min(...xs) - pad;
  const y = Math.min(...ys) - pad;
  const w = Math.max(...xs) - Math.min(...xs) + pad * 2;
  const h = Math.max(...ys) - Math.min(...ys) + pad * 2;
  return { x, y, w: Math.max(w, 2), h: Math.max(h, 2) };
}

/** Page-point gesture → the normalised payload the API and the PDF read. */
export function toDrawingOptions(
  points: [number, number][],
  box: { x: number; y: number; w: number; h: number },
  style: { color: string; stroke: number },
): DrawingOptions {
  const round = (n: number) => Math.round(Math.max(0, Math.min(1, n)) * 10000) / 10000;
  return {
    kind: 'drawing',
    color: color(style.color),
    stroke: clamp(style.stroke, MIN_PEN_WIDTH, MAX_PEN_WIDTH, DEFAULT_PEN_WIDTH),
    strokes: [points.map(([px, py]) => [round((px - box.x) / box.w), round((py - box.y) / box.h)] as [number, number])],
  };
}
